### PR TITLE
Feat convert send nitro

### PR DESCRIPTION
**Pull request by gtokman**

### CODE REVIEW

File: packages/npm/send/.yarnrc.yml [hunk 1]
```diff
@@ -1,4 +1,6 @@
-enableGlobalCache: true
+compressionLevel: mixed
```
Comment (liambutler-lawrence) [inline packages/npm/send/.yarnrc.yml:1]: What does this do?

Comment (gtokman) [reply packages/npm/send/.yarnrc.yml:1]: https://yarnpkg.com/configuration/yarnrc

File: packages/npm/send/CODE_OF_CONDUCT.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+an individual is officially representing the community in public spaces.
+Examples of representing our community include using an official e-mail address,
+posting via an official social media account, or acting as an appointed
+representative at an online or offline event.
+
+## Enforcement
+
+Instances of abusive, harassing, or otherwise unacceptable behavior may be
+reported to the community leaders responsible for enforcement at
+[INSERT CONTACT METHOD].
```
Comment (liambutler-lawrence) [inline packages/npm/send/CODE_OF_CONDUCT.md:64]: ?

File: packages/npm/send/CODE_OF_CONDUCT.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
+
```
Comment (liambutler-lawrence) [inline packages/npm/send/CODE_OF_CONDUCT.md:1]: We already have this file at root, why duplicate it here?

File: packages/npm/send/LICENSE [hunk 4]
```diff
@@ -0,0 +1,20 @@
+MIT License
```
Comment (liambutler-lawrence) [inline packages/npm/send/LICENSE:1]: We already have this file at root, why duplicate it here?

File: packages/npm/send/example/ios/File.swift [hunk 5]
```diff
@@ -0,0 +1,6 @@
+//
```
Comment (liambutler-lawrence) [inline packages/npm/send/example/ios/File.swift:1]: What is this for?

File: packages/npm/send/example/ios/Podfile [hunk 6]
```diff
@@ -19,8 +21,7 @@ target 'SendExample' do
 
   use_react_native!(
     :path => config[:reactNativePath],
-    :hermes_enabled => false,
-    :fabric_enabled => false,
```
Comment (liambutler-lawrence) [inline packages/npm/send/example/ios/Podfile:23]: Does this mean we're using Hermes now? And whatever Fabric is?

File: packages/npm/send/example/ios/Podfile.lock [hunk 7]
```diff
@@ -1302,71 +1733,86 @@ EXTERNAL SOURCES:
...
+  ReactCommon: 6ef348087d250257c44c0204461c03f036650e9b
   SocketRocket: abac6f5de4d4d62d24e11868d7a2f427e0ef940d
-  Yoga: 56f906bf6c11c931588191dde1229fd3e4e3d557
+  Yoga: 2a45d7e59592db061217551fd3bbe2dd993817ae
 
-PODFILE CHECKSUM: 62a1950978e21c567801b660edb3d9841ca98967
+PODFILE CHECKSUM: 587829e2504f7c0b375fddc67c0ec17f3b28a967
 
-COCOAPODS: 1.14.3
+COCOAPODS: 1.15.2
```
Comment (liambutler-lawrence) [inline packages/npm/send/example/ios/Podfile.lock:1818]: Please re-run with `bundle exec`

File: packages/npm/send/example/ios/SendExample.xcodeproj/xcshareddata/xcschemes/SendExample.xcscheme [hunk 8]
```diff
@@ -5,6 +5,15 @@
    <BuildAction
       parallelizeBuildables = "YES"
       buildImplicitDependencies = "YES">
+      <PreActions>
+         <ExecutionAction
+            ActionType = "Xcode.IDEStandardExecutionActionsCore.ExecutionActionType.ShellScriptAction">
+            <ActionContent
+               title = "Invoke Codegen"
+               scriptText = "#cd &quot;$WORKSPACE_PATH/../../../&quot; &amp;&amp; npx bob build --target codegen&#10;">
```
Comment (liambutler-lawrence) [inline packages/npm/send/example/ios/SendExample.xcodeproj/xcshareddata/xcschemes/SendExample.xcscheme:13]: Why do we need to do this both on pod install and on build? Shouldn't one be sufficient?

File: packages/npm/send/lefthook.yml [hunk 9]
```diff
@@ -6,7 +6,7 @@ pre-commit:
       run: npx eslint {staged_files}
     types:
       glob: "*.{js,ts, jsx, tsx}"
-      run: npx tsc --noEmit
```
Comment (liambutler-lawrence) [inline packages/npm/send/lefthook.yml:9]: Why this change?

File: packages/npm/send/src/index.tsx [hunk 10]
```diff
@@ -1,50 +1,22 @@
...
-}
-
-export async function send(request: Request): Promise<Response> {
-  // NOTE: The React Native bridge stringifies all objects anyway; we just do it manually so we can take advantage of Swift.Codable and kotlinx.serialization to simplify the native code
-  const stringifiedRequest = JSON.stringify(request)
-  const stringifiedResponse = await Send.send(stringifiedRequest)
-  return JSON.parse(stringifiedResponse)
+export async function send(
+  request: Request & { method: Method }
+): Promise<SendResult> {
```
Comment (liambutler-lawrence) [inline packages/npm/send/src/index.tsx:20]:
```suggestion
): Promise<(Response & { result: 'success' }) | (SendError & { result: 'error' })> {
```

We should be exposing the public interface as a proper discriminated union regardless of what hack we have to use for the internal JSI interface

File: packages/npm/send/src/index.tsx [hunk 11]
```diff
@@ -1,50 +1,22 @@
...
+  return Send.send(request)
```
Comment (liambutler-lawrence) [inline packages/npm/send/src/index.tsx:21]: We should wrap this in a try/catch to guarantee all errors will be returned via the result type

File: packages/npm/send/src/index.tsx [hunk 12]
```diff
@@ -1,50 +1,22 @@
...
-  body: string | null
-}
-
-export async function send(request: Request): Promise<Response> {
-  // NOTE: The React Native bridge stringifies all objects anyway; we just do it manually so we can take advantage of Swift.Codable and kotlinx.serialization to simplify the native code
-  const stringifiedRequest = JSON.stringify(request)
-  const stringifiedResponse = await Send.send(stringifiedRequest)
-  return JSON.parse(stringifiedResponse)
+export async function send(
+  request: Request & { method: Method }
```
Comment (liambutler-lawrence) [inline packages/npm/send/src/index.tsx:19]: Nitro folks already fixed conversion of string enums, so we should be able to use this in the actual JSI interface now https://github.com/mrousavy/nitro/pull/87

File: packages/npm/send/src/index.tsx [hunk 13]
```diff
@@ -1,50 +1,22 @@
-import { NativeModules, Platform } from 'react-native'
+import { NitroModules } from 'react-native-nitro-modules'
+import type { Request, Send, SendResult } from './Send.nitro'
+export * from './Send.nitro'
```
Comment (liambutler-lawrence) [inline packages/npm/send/src/index.tsx:3]:
```suggestion
import type { Request as _Request, Send } from './Send.nitro'

export type Request = _Request & { method: Method }
export type SendResult = (Response & { result: 'success' }) | (SendError & { result: 'error' })
export { Parameters, Response, Code } from './Send.nitro'
```

We don't want to expose SendResult or Send as those are internal to the JSI interface.

File: packages/npm/send/src/index.tsx (hunk 12 above)
Comment (liambutler-lawrence) [inline packages/npm/send/src/index.tsx:19]:
```suggestion
  request: Request
```

Redefined Request above

File: packages/npm/send/src/index.tsx (hunk 13 above)
Comment (liambutler-lawrence) [reply packages/npm/send/src/index.tsx:3]: We also want to redefine the Request and SendResult types so we don't expose the internal/hacked versions

File: packages/npm/send/src/Send.nitro.ts [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+export interface SendError {
+  code: Code
+  message: string
+}
+
+export interface SendResult {
+  response?: Response
+  error?: SendError
```
Comment (liambutler-lawrence) [inline packages/npm/send/src/Send.nitro.ts:44]: We should be consistent throughout this interface, either use `null` or `undefined` but not both

File: packages/npm/send/src/Send.nitro.ts [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+enum Code {
+  RESPONSE_INVALID,
+  REQUEST_INVALID,
+  NETWORK_ERROR,
+  NO_RESPONSE,
+  NON_UTF8_RESPONSE_BODY,
+  NON_UTF8_REQUEST_BODY,
+  INVALID_REQUEST_PATH_OR_QUERY_PARAMETERS,
+  IVNALID_REQUEST_BASE_URL,
```
Comment (liambutler-lawrence) [inline packages/npm/send/src/Send.nitro.ts:31]:
```suggestion
  INVALID_REQUEST_BASE_URL,
```

File: packages/npm/send/ios/Send.swift [hunk 16]
```diff
@@ -18,78 +8,29 @@ private func bodyIsUTF8(contentTypeHeader: String?, utf8ContentTypes: [String])
...
-@propertyWrapper
-struct RequiredKey<Value> {
-    var wrappedValue: Value?
-}
-
-extension RequiredKey: Decodable where Value: Decodable {
-    init(from decoder: Decoder) throws {
-        let container = try decoder.singleValueContainer()
-        wrappedValue = container.decodeNil() ? nil : try container.decode(Value.self)
+final class IgnoreRedirectsDelegate: NSObject, URLSessionTaskDelegate {
```
Comment (liambutler-lawrence) [inline packages/npm/send/ios/Send.swift:11]: Can you move this back where it was to avoid the unnecessary diff?

File: packages/npm/send/ios/Send.swift [hunk 17]
```diff
@@ -18,78 +8,29 @@ private func bodyIsUTF8(contentTypeHeader: String?, utf8ContentTypes: [String])
...
-    let headerParameters: [String: String]
-    let method: String
-    @RequiredKey var body: String?
-    let utf8ContentTypes: [String]
-    
+extension Request {
     var url: Result<URL, SendError> {
         guard var urlComponents = URLComponents(string: baseURL) else {
-            return .failure(.invalidRequestBaseURL)
+            return .failure(SendError(code: .ivnalid_request_base_url, message: ""))
```
Comment (liambutler-lawrence) [inline packages/npm/send/ios/Send.swift:22]:
```suggestion
            return .failure(SendError(code: .ivnalid_request_base_url, message: ""))
```

We should add back the human-readable strings...

File: packages/npm/send/ios/Send.swift [hunk 18]
```diff
@@ -18,78 +8,29 @@ private func bodyIsUTF8(contentTypeHeader: String?, utf8ContentTypes: [String])
...
-            urlComponents.queryItems = queryParameters.map { key, value in
+        if !query.parameters.isEmpty {
+            urlComponents.queryItems = query.parameters.map { key, value in
                 URLQueryItem(name: key, value: value)
             }
         }
         
         guard let url = urlComponents.url else {
-            return .failure(.invalidRequestPathOrQueryParameters)
+            return .failure(SendError(code: .invalid_request_path_or_query_parameters, message: ""))
```
Comment (liambutler-lawrence) [inline packages/npm/send/ios/Send.swift:33]: Same here

File: packages/npm/send/ios/Send.swift [hunk 19]
```diff
@@ -99,20 +40,20 @@ struct Request: Decodable {
...
                 urlRequest.setValue(value, forHTTPHeaderField: key)
             }
             
             if let body {
-                let contentTypeHeader = headerParameters.first(where: { $0.key.caseInsensitiveCompare("Content-Type") == .orderedSame })?.value
+                let contentTypeHeader = header.parameters.first(where: { $0.key.caseInsensitiveCompare("Content-Type") == .orderedSame })?.value
                 if bodyIsUTF8(contentTypeHeader: contentTypeHeader, utf8ContentTypes: utf8ContentTypes) {
                     guard let utf8Body = body.data(using: .utf8) else {
-                        return .failure(.nonUTF8RequestBody)
+                        return .failure(SendError(code: .non_utf8_request_body, message: ""))
```
Comment (liambutler-lawrence) [inline packages/npm/send/ios/Send.swift:51]: And here

File: packages/npm/send/ios/Send.swift [hunk 20]
```diff
@@ -99,20 +40,20 @@ struct Request: Decodable {
...
                     }
                     urlRequest.httpBody = utf8Body
                 } else {
                     guard let base64Body = Data(base64Encoded: body) else {
-                        return .failure(.nonBase64RequestBody)
+                        return .failure(SendError(code: .non_base64_request_body, message: ""))
```
Comment (liambutler-lawrence) [inline packages/npm/send/ios/Send.swift:56]: And here

File: packages/npm/send/ios/Send.swift [hunk 21]
```diff
@@ -122,160 +63,144 @@ struct Request: Decodable {
...
-    let headerParameters: [String: String]
-    @RequiredKey var body: String?
-    
+extension Response {
     init(request: Request, data: Data, httpURLResponse: HTTPURLResponse) throws {
-        self.statusCode = httpURLResponse.statusCode
-        
         guard let headerParameters = httpURLResponse.allHeaderFields as? [String: String] else {
-            throw SendError.invalidResponseHeaderParameters
+            throw SendError(code: .invalid_response_header_parameters, message: "")
```
Comment (liambutler-lawrence) [inline packages/npm/send/ios/Send.swift:69]: And here

File: packages/npm/send/ios/Send.swift [hunk 22]
```diff
@@ -122,160 +63,144 @@ struct Request: Decodable {
...
+            body = nil
         } else {
             let bodyIsUTF8 = bodyIsUTF8(
                 contentTypeHeader: httpURLResponse.value(forHTTPHeaderField: "Content-Type"),
                 utf8ContentTypes: request.utf8ContentTypes
             )
             if (bodyIsUTF8) {
                 guard let utf8Body = String(data: data, encoding: .utf8) else {
-                    throw SendError.nonUTF8ResponseBody
+                    throw SendError(code: .non_utf8_response_body, message: "")
```
Comment (liambutler-lawrence) [inline packages/npm/send/ios/Send.swift:81]: And here

File: packages/npm/send/ios/Send.swift [hunk 23]
```diff
@@ -122,160 +63,144 @@ struct Request: Decodable {
...
+        )
     }
 }
 
-final class IgnoreRedirectsDelegate: NSObject, URLSessionTaskDelegate {
-    func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse, newRequest request: URLRequest) async -> URLRequest? {
-        return nil
-    }
+
+final class NetworkManager {
```
Comment (liambutler-lawrence) [inline packages/npm/send/ios/Send.swift:97]: Why make a whole separate class for this?

File: packages/npm/send/ios/Send.swift [hunk 24]
```diff
@@ -122,160 +63,144 @@ struct Request: Decodable {
...
-                        .downloadDecodingFailedToComplete,
-                        .httpTooManyRedirects,
-                        .redirectToNonExistentLocation,
-                        .zeroByteResource:
-                    reject(CODE_RESPONSE_INVALID, MESSAGE_RESPONSE_INVALID, urlError)
-                    
-                default:
-                    // NOTE: The only other documented case is `unknown`, but code is not an enum so a default case is required regardless
-                    reject(CODE_UNKNOWN, MESSAGE_UNKNOWN, urlError)
+    var hybridContext = margelo.nitro.HybridContext()
```
Comment (liambutler-lawrence) [inline packages/npm/send/ios/Send.swift:113]:
```suggestion
    let hybridContext = margelo.nitro.HybridContext()
```

File: packages/npm/send/example/ios/Podfile.lock (hunk 7 above)
Comment (gtokman) [reply packages/npm/send/example/ios/Podfile.lock:1818]: this is running bundle exec, I upgraded the project

File: packages/npm/send/example/ios/SendExample.xcodeproj/xcshareddata/xcschemes/SendExample.xcscheme (hunk 8 above)
Comment (gtokman) [reply packages/npm/send/example/ios/SendExample.xcodeproj/xcshareddata/xcschemes/SendExample.xcscheme:13]: it's commented out

File: packages/npm/send/example/ios/Podfile (hunk 6 above)
Comment (gtokman) [reply packages/npm/send/example/ios/Podfile:23]: just the example app lol

File: packages/npm/send/example/ios/File.swift (hunk 5 above)
Comment (gtokman) [reply packages/npm/send/example/ios/File.swift:1]: bridging header

File: packages/npm/send/LICENSE (hunk 4 above)
Comment (gtokman) [reply packages/npm/send/LICENSE:1]: prob should have a license for every package

File: packages/npm/send/ios/Send.swift (hunk 23 above)
Comment (gtokman) [reply packages/npm/send/ios/Send.swift:97]: So that it's only created once like URLSession.shared

Comment (liambutler-lawrence) [reply packages/npm/send/ios/Send.swift:97]: But the Send class is only created once anyway...

File: packages/npm/send/ios/Send.swift [hunk 25]
```diff
@@ -122,160 +63,144 @@ struct Request: Decodable {
...
+    lazy var session = NetworkManager.shared.session
+    
+    func send(request: Request) throws -> NitroModules.Promise<SendResult> {
+        return Promise.async { [weak self] in
+            if let self {
+                do {
+                    let urlRequest = try request.urlRequest.get()
+                    let (data, urlResponse) = try await session.data(for: urlRequest)
+                    guard let httpURLResponse = urlResponse as? HTTPURLResponse else {
+                        throw SendError(code: .non_utf8_response_body, message: "")
```
Comment (liambutler-lawrence) [inline packages/npm/send/ios/Send.swift:128]: And here

File: packages/npm/send/ios/Send.swift [hunk 26]
```diff
@@ -122,160 +63,144 @@ struct Request: Decodable {
...
-                    reject(CODE_UNKNOWN, MESSAGE_UNKNOWN, urlError)
+    var hybridContext = margelo.nitro.HybridContext()
+    
+    var memorySize: Int {
+        return getSizeOf(self)
+    }
+    
+    lazy var session = NetworkManager.shared.session
+    
+    func send(request: Request) throws -> NitroModules.Promise<SendResult> {
```
Comment (liambutler-lawrence) [inline packages/npm/send/ios/Send.swift:121]:
```suggestion
    func send(request: Request) -> NitroModules.Promise<SendResult> {
```

File: packages/npm/send/ios/Send.swift (hunk 24 above)
Comment (gtokman) [reply packages/npm/send/ios/Send.swift:113]: has to be a var that's how it's defined in the protocol

File: packages/npm/send/ios/Send.swift (hunk 25 above)
Comment (liambutler-lawrence) [reply packages/npm/send/ios/Send.swift:128]: Might as well just return from here, no point in throwing to catch it a few lines below

File: packages/npm/send/ios/Send.swift [hunk 27]
```diff
@@ -122,160 +63,144 @@ struct Request: Decodable {
...
+                            .dataNotAllowed,
+                            .dnsLookupFailed,
+                            .fileDoesNotExist,
+                            .fileIsDirectory,
+                            .noPermissionsToReadFile,
+                            .requestBodyStreamExhausted,
+                            .unsupportedURL,
+                            .userAuthenticationRequired,
+                            .userCancelledAuthentication:
+                        let sendError = SendError(code: .response_invalid, message: urlError.localizedDescription)
```
Comment (liambutler-lawrence) [inline packages/npm/send/ios/Send.swift:157]:
```suggestion
                        let sendError = SendError(code: .request_invalid, message: urlError.localizedDescription)
```

File: packages/npm/send/ios/Send.swift (hunk 25 above)
Comment (liambutler-lawrence) [inline packages/npm/send/ios/Send.swift:128]:
```suggestion
                        throw SendError(code: .response_invalid, message: "")
```

File: packages/npm/send/ios/Send.swift [hunk 28]
```diff
@@ -122,160 +63,144 @@ struct Request: Decodable {
...
-                        .cannotWriteToFile,
-                        .downloadDecodingFailedMidStream,
-                        .downloadDecodingFailedToComplete,
-                        .httpTooManyRedirects,
-                        .redirectToNonExistentLocation,
-                        .zeroByteResource:
-                    reject(CODE_RESPONSE_INVALID, MESSAGE_RESPONSE_INVALID, urlError)
-                    
-                default:
-                    // NOTE: The only other documented case is `unknown`, but code is not an enum so a default case is required regardless
```
Comment (liambutler-lawrence) [inline packages/npm/send/ios/Send.swift:274]: Why remove this comment?

File: packages/npm/send/ios/Send.swift [hunk 29]
```diff
@@ -122,160 +63,144 @@ struct Request: Decodable {
...
+                            .cannotWriteToFile,
+                            .downloadDecodingFailedMidStream,
+                            .downloadDecodingFailedToComplete,
+                            .httpTooManyRedirects,
+                            .redirectToNonExistentLocation,
+                            .zeroByteResource:
+                        let sendError = SendError(code: .response_invalid, message: urlError.localizedDescription)
+                        return SendResult(response: nil, error: sendError)
+                    default:
+                        let sendError = SendError(code: .unexpected, message: "Something went wrong. Please file an issue on GitHub or try again.")
```
Comment (liambutler-lawrence) [inline packages/npm/send/ios/Send.swift:194]: Why not use the error.localizedDescription like all the other cases?

File: packages/npm/send/ios/Send.swift [hunk 30]
```diff
@@ -122,160 +63,144 @@ struct Request: Decodable {
...
+    var hybridContext = margelo.nitro.HybridContext()
+    
+    var memorySize: Int {
+        return getSizeOf(self)
+    }
+    
+    lazy var session = NetworkManager.shared.session
+    
+    func send(request: Request) throws -> NitroModules.Promise<SendResult> {
+        return Promise.async { [weak self] in
```
Comment (liambutler-lawrence) [inline packages/npm/send/ios/Send.swift:122]: Is adding [weak self] necessary? If it is, I would make the copy returned in the error case a little more specific (eg `Nitro module deinitialized.`)

File: packages/npm/send/ios/NitroSendOnLoad.mm [hunk 31]
```diff
@@ -0,0 +1,27 @@
+#import "HybridSendSpecSwift.hpp"
+#import "HybridSendSpec.hpp"
+#import <NitroModules/HybridObjectRegistry.hpp>
+#import <Foundation/Foundation.h>
+
+namespace NitroSend {
+class NitroSendRegistry;
+} // namespace NitroImage
```
Comment (liambutler-lawrence) [inline packages/npm/send/ios/NitroSendOnLoad.mm:8]:
```suggestion
}
```

File: packages/npm/send/ios/Send.mm [hunk 32]
```diff
@@ -1,14 +0,0 @@
-#import <React/RCTBridgeModule.h>
-
-@interface RCT_EXTERN_MODULE(Send, NSObject)
-
-RCT_EXTERN_METHOD(send:(NSString *)request
-                  withResolver:(RCTPromiseResolveBlock)resolve
-                  withRejecter:(RCTPromiseRejectBlock)reject)
-
-+ (BOOL)requiresMainQueueSetup
```
Comment (liambutler-lawrence) [inline packages/npm/send/ios/Send.mm:9]: Where is this flag set now?

File: packages/npm/send/example/Gemfile [hunk 33]
```diff
@@ -3,7 +3,6 @@ source 'https://rubygems.org'
 # You may use http://rbenv.org/ or https://rvm.io/ to install and use this version
 ruby ">= 2.6.10"
 
-# Cocoapods 1.15 introduced a bug which break the build. We will remove the upper
-# bound in the template on Cocoapods with next React Native release.
-gem 'cocoapods', '>= 1.13', '< 1.15'
-gem 'activesupport', '>= 6.1.7.5', '< 7.1.0'
+# Exclude problematic versions of cocoapods and activesupport that causes build failures.
+gem 'cocoapods', '>= 1.13', '!= 1.15.0', '!= 1.15.1'
```
Comment (liambutler-lawrence) [inline packages/npm/send/example/Gemfile:7]:
```suggestion
gem 'cocoapods', '>= 1.15.2'
```
Wouldn't this be simpler? Same for activesupport below

File: packages/npm/send/example/ios/File.swift (hunk 5 above)
Comment (liambutler-lawrence) [reply packages/npm/send/example/ios/File.swift:1]: Say more... if it's actually necessary to have a dummy file like this, at least leave a `// NOTE: This is necessary because` comment at the top

File: packages/npm/send/example/ios/Podfile.lock (hunk 7 above)
Comment (liambutler-lawrence) [reply packages/npm/send/example/ios/Podfile.lock:1818]: My bad, just saw that!

File: packages/npm/send/example/src/App.tsx [hunk 34]
```diff
@@ -15,52 +23,110 @@ export const serializeError = (error: unknown) => {
...
-        'Content-Type': 'application/json',
-        'Accept': 'application/json',
+      header: {
+        parameters: {
+          'Cache-Control': 'no-cache',
+          'Content-Type': 'application/json',
+          'Accept': 'application/json',
+        },
       },
+    }).then((result) => {
```
Comment (liambutler-lawrence) [inline packages/npm/send/example/src/App.tsx:56]: #nth use async/await

File: packages/npm/send/example/src/App.tsx [hunk 35]
```diff
@@ -15,52 +23,110 @@ export const serializeError = (error: unknown) => {
...
+      if (result.error) {
+        console.log('ERROR', serializeError(result.error))
+      } else {
+        setData(JSON.parse(result.response?.body as any).results[0])
```
Comment (liambutler-lawrence) [inline packages/npm/send/example/src/App.tsx:60]: This is exactly why we need the `send` function to return a discriminated union (regardless of what the internal JSI interface uses)

File: packages/npm/send/example/src/App.tsx [hunk 36]
```diff
@@ -15,52 +23,110 @@ export const serializeError = (error: unknown) => {
...
-        console.log('SUCCESS', response)
-        setResult(response.body ?? 'no body')
-      })
-      .catch((error) => {
-        console.log('FAILURE', serializeError(error))
-        setResult(`[${error.code}] ${error.message}`)
-      })
   }, [])
 
+  if (!data) {
```
Comment (liambutler-lawrence) [inline packages/npm/send/example/src/App.tsx:65]:
```suggestion
  if (data === null) {
```
Falsiness/truthiness checks in TS are overly broad, have unintuitive behavior, and should not be used

File: packages/npm/send/example/src/App.tsx [hunk 37]
```diff
@@ -15,52 +23,110 @@ export const serializeError = (error: unknown) => {
...
-  const [result, setResult] = React.useState<string>('waiting...')
+const App = () => {
+  const [data, setData] = useState<{
+    trackName: string
+    artistName: string
+    artworkUrl100: string
+    description: string
+    version: string
+    screenshotUrls: string[]
+  } | null>()
```
Comment (liambutler-lawrence) [inline packages/npm/send/example/src/App.tsx:34]:
```suggestion
  } | null>(null)
```
Otherwise, won't this default to undefined?

File: packages/npm/send/example/src/App.tsx [hunk 38]
```diff
@@ -1,8 +1,16 @@
...
+  Image,
+  SafeAreaView,
+  ScrollView,
+  StyleSheet,
+  Text,
+  View,
+} from 'react-native'
 
-export const serializeError = (error: unknown) => {
+const serializeError = (error: unknown) => {
```
Comment (liambutler-lawrence) [inline packages/npm/send/example/src/App.tsx:12]: We don't need this function anymore... the whole point of returning a `SendError` is that it's type-safe and has a defined schema

File: packages/npm/send/ios/Send.mm (hunk 32 above)
Comment (gtokman) [reply packages/npm/send/ios/Send.mm:9]: not sure, guessing it's not needed in JSI.

File: packages/npm/send/ios/NitroSendOnLoad.mm (hunk 31 above)
Comment (gtokman) [reply packages/npm/send/ios/NitroSendOnLoad.mm:8]: this is a cpp convention.

File: packages/npm/send/ios/Send.swift (hunk 23 above)
Comment (gtokman) [reply packages/npm/send/ios/Send.swift:97]: I'm not 100% certain that is true, are you?

File: packages/npm/send/example/ios/File.swift (hunk 5 above)
Comment (gtokman) [reply packages/npm/send/example/ios/File.swift:1]: it was generated by the package create package

File: packages/npm/send/ios/Send.swift (hunk 24 above)
Comment (liambutler-lawrence) [reply packages/npm/send/ios/Send.swift:112]: Sad, will complain about that in Discord lol

File: packages/npm/send/ios/Send.swift [hunk 39]
```diff
@@ -18,101 +8,46 @@ private func bodyIsUTF8(contentTypeHeader: String?, utf8ContentTypes: [String])
...
                 if bodyIsUTF8(contentTypeHeader: contentTypeHeader, utf8ContentTypes: utf8ContentTypes) {
                     guard let utf8Body = body.data(using: .utf8) else {
-                        return .failure(.nonUTF8RequestBody)
+                        return .failure(SendError(code: .nonUtf8RequestBody, message: "Your request headers specify a Content-Type included in `utf8ContentTypes`, but your request body is not a UTF8-encoded string."))
                     }
                     urlRequest.httpBody = utf8Body
                 } else {
                     guard let base64Body = Data(base64Encoded: body) else {
-                        return .failure(.nonBase64RequestBody)
+                        return .failure(SendError(code: .nonUtf8RequestBody, message: "Your request headers specify a Content-Type NOT included in `utf8ContentTypes`, but your request body is not a base64-encoded string."))
```
Comment (liambutler-lawrence) [inline packages/npm/send/ios/Send.swift:50]:
```suggestion
                        return .failure(SendError(code: .nonBase64RequestBody, message: "Your request headers specify a Content-Type NOT included in `utf8ContentTypes`, but your request body is not a base64-encoded string."))
```

Regression from nitro upgrade commit

**Review by liambutler-lawrence — APPROVED** [review]
